server/src/com/thoughtworks/go/config/CachedFileGoConfig.java

```diff
@@ -95,9 +96,12 @@ public void onTimer() {
 
     private synchronized void loadFromDisk() {
```

Comment:
Technically since all entry points that we have today for config saves are synchronized on the same lock, we shouldn't see more deadlocks. For future entry-points, we just need to ensure that the lock is taken before any config save operation is done. 
Having said that, you are right that we do not need so many `synchronized` methods on some of these classes. I was a little jittery about removing them. I wanted `ConfigSaveDeadlockDetectionIntegrationTest` to be able to test such scenarios, but wanted to be doubly sure that I haven't missed any scenario in that. If you think that is good enough test, then after removing the `synchronized` blocks, the methods look something like this:

``` java
diff --git a/server/src/com/thoughtworks/go/config/CachedFileGoConfig.java b/server/src/com/thoughtworks/go/config/CachedFileGoConfig.java
index 20284ca..0ad3671 100644
--- a/server/src/com/thoughtworks/go/config/CachedFileGoConfig.java
+++ b/server/src/com/thoughtworks/go/config/CachedFileGoConfig.java
@@ -131,23 +131,23 @@ public class CachedFileGoConfig implements CachedGoConfig {
     }
 
     @Override
-    public synchronized ConfigSaveState writeWithLock(UpdateConfigCommand updateConfigCommand) {
+    public ConfigSaveState writeWithLock(UpdateConfigCommand updateConfigCommand) {
         GoConfigHolder holder = new GoConfigHolder(currentConfig, currentConfigForEdit);
         return writeWithLock(updateConfigCommand, holder);
     }
 
-    public synchronized ConfigSaveState writeWithLock(UpdateConfigCommand updateConfigCommand, GoConfigHolder holder) {
+    public ConfigSaveState writeWithLock(UpdateConfigCommand updateConfigCommand, GoConfigHolder holder) {
         GoFileConfigDataSource.GoConfigSaveResult saveResult = dataSource.writeWithLock(updateConfigCommand, holder);
         saveValidConfigToCacheAndNotifyConfigChangeListeners(saveResult.getConfigHolder());
         return saveResult.getConfigSaveState();
     }
 
-    public synchronized void writePipelineWithLock(PipelineConfig pipelineConfig, PipelineConfigService.SaveCommand saveCommand, Username currentUser) {
+    public void writePipelineWithLock(PipelineConfig pipelineConfig, PipelineConfigService.SaveCommand saveCommand, Username currentUser) {
         GoConfigHolder serverCopy = new GoConfigHolder(currentConfig, currentConfigForEdit);
         writePipelineWithLock(pipelineConfig, serverCopy, saveCommand, currentUser);
     }
 
-    public synchronized PipelineConfigSaveResult writePipelineWithLock(PipelineConfig pipelineConfig, GoConfigHolder serverCopy, PipelineConfigService.SaveCommand saveCommand, Username currentUser) {
+    public PipelineConfigSaveResult writePipelineWithLock(PipelineConfig pipelineConfig, GoConfigHolder serverCopy, PipelineConfigService.SaveCommand saveCommand, Username currentUser) {
         PipelineConfigSaveResult saveResult = dataSource.writePipelineWithLock(pipelineConfig, serverCopy, saveCommand, currentUser);
         saveValidConfigToCacheAndNotifyPipelineConfigChangeListeners(saveResult);
         return saveResult;
@@ -170,14 +170,14 @@ public class CachedFileGoConfig implements CachedGoConfig {
         LOGGER.info("Finished notifying pipeline config listeners");
     }
 
-    private synchronized void saveValidConfigToCacheAndNotifyConfigChangeListeners(GoConfigHolder configHolder) {
+    private void saveValidConfigToCacheAndNotifyConfigChangeListeners(GoConfigHolder configHolder) {
         saveValidConfigToCache(configHolder);
         if(configHolder!=null) {
             notifyListeners(currentConfig);
         }
     }
 
-    private synchronized void saveValidConfigToCache(GoConfigHolder configHolder) {
+    private void saveValidConfigToCache(GoConfigHolder configHolder) {
         if (configHolder != null) {
             LOGGER.debug("[Config Save] Saving config to the cache");
             this.lastException = null;
@@ -204,7 +204,7 @@ public class CachedFileGoConfig implements CachedGoConfig {
     }
 
     @Override
-    public synchronized void save(String configFileContent, boolean shouldMigrate) throws Exception {
+    public void save(String configFileContent, boolean shouldMigrate) throws Exception {
         GoConfigHolder newConfigHolder = dataSource.write(configFileContent, shouldMigrate);
         saveValidConfigToCacheAndNotifyConfigChangeListeners(newConfigHolder);
     }
@@ -226,7 +226,7 @@ public class CachedFileGoConfig implements CachedGoConfig {
         }
     }
 
-    private synchronized void notifyListeners(CruiseConfig newCruiseConfig) {
+    private void notifyListeners(CruiseConfig newCruiseConfig) {
         LOGGER.info("About to notify config listeners");
         for (ConfigChangedListener listener : listeners) {
             try {
diff --git a/server/src/com/thoughtworks/go/config/GoConfigWatchList.java b/server/src/com/thoughtworks/go/config/GoConfigWatchList.java
index 3fbbc05..b9fdb34 100644
--- a/server/src/com/thoughtworks/go/config/GoConfigWatchList.java
+++ b/server/src/com/thoughtworks/go/config/GoConfigWatchList.java
@@ -57,7 +57,7 @@ public class GoConfigWatchList implements ConfigChangedListener {
         this.reposConfig = partSources;
         notifyListeners(partSources);
     }
-    private synchronized void notifyListeners(ConfigReposConfig configRepoConfigs) {
+    private void notifyListeners(ConfigReposConfig configRepoConfigs) {
         for (ChangedRepoConfigWatchListListener listener : listeners) {
             try {
                 listener.onChangedRepoConfigWatchList(configRepoConfigs);
diff --git a/server/src/com/thoughtworks/go/config/GoConfigWriteLock.java b/server/src/com/thoughtworks/go/config/GoConfigWriteLock.java
index 4ed3ae3..98750f2 100644
--- a/server/src/com/thoughtworks/go/config/GoConfigWriteLock.java
+++ b/server/src/com/thoughtworks/go/config/GoConfigWriteLock.java
@@ -20,13 +20,4 @@ import org.springframework.stereotype.Component;
 
 @Component
 public class GoConfigWriteLock {
-    private final Object writeLock;
-
-    public GoConfigWriteLock() {
-        this.writeLock = new Object();
-    }
-
-    public Object mutex() {
-        return writeLock;
-    }
 }
diff --git a/server/src/com/thoughtworks/go/config/GoFileConfigDataSource.java b/server/src/com/thoughtworks/go/config/GoFileConfigDataSource.java
index 7ccc67c..4c3635b 100644
--- a/server/src/com/thoughtworks/go/config/GoFileConfigDataSource.java
+++ b/server/src/com/thoughtworks/go/config/GoFileConfigDataSource.java
@@ -140,7 +140,7 @@ public class GoFileConfigDataSource {
         }
     }
 
-    synchronized GoConfigHolder forceLoad(File configFile) throws Exception {
+    GoConfigHolder forceLoad(File configFile) throws Exception {
         if (LOGGER.isDebugEnabled()) {
             LOGGER.debug("Reloading config file: " + configFile.getAbsolutePath());
         }
@@ -160,7 +160,7 @@ public class GoFileConfigDataSource {
     }
 
     @Deprecated
-    public synchronized GoConfigHolder write(String configFileContent, boolean shouldMigrate) throws Exception {
+    public GoConfigHolder write(String configFileContent, boolean shouldMigrate) throws Exception {
         File configFile = fileLocation();
         FileOutputStream output = null;
         try {
@@ -211,7 +211,7 @@ public class GoFileConfigDataSource {
         }
     }
 
-    public synchronized CachedFileGoConfig.PipelineConfigSaveResult writePipelineWithLock(PipelineConfig pipelineConfig, GoConfigHolder serverCopy, PipelineConfigService.SaveCommand saveCommand, Username currentUser) {
+    public CachedFileGoConfig.PipelineConfigSaveResult writePipelineWithLock(PipelineConfig pipelineConfig, GoConfigHolder serverCopy, PipelineConfigService.SaveCommand saveCommand, Username currentUser) {
         CruiseConfig modifiedConfig = cloner.deepClone(serverCopy.configForEdit);
         saveCommand.updateConfig(modifiedConfig, pipelineConfig);
         CruiseConfig preprocessedConfig = cloner.deepClone(modifiedConfig);
@@ -237,7 +237,7 @@ public class GoFileConfigDataSource {
         }
     }
 
-    public synchronized GoConfigSaveResult writeWithLock(UpdateConfigCommand updatingCommand, GoConfigHolder configHolder) {
+    public GoConfigSaveResult writeWithLock(UpdateConfigCommand updatingCommand, GoConfigHolder configHolder) {
         try {
 
             // Need to convert to xml before we try to write it to the config file.
diff --git a/server/src/com/thoughtworks/go/config/MergedGoConfig.java b/server/src/com/thoughtworks/go/config/MergedGoConfig.java
index 4dadc54..ca06859 100644
--- a/server/src/com/thoughtworks/go/config/MergedGoConfig.java
+++ b/server/src/com/thoughtworks/go/config/MergedGoConfig.java
@@ -141,12 +141,12 @@ public class MergedGoConfig implements CachedGoConfig, ConfigChangedListener, Pa
 
     // no actions on timer now. We only react to events in CachedFileGoConfig and in GoPartialConfig
 
-    public synchronized ConfigSaveState writeWithLock(UpdateConfigCommand updateConfigCommand) {
+    public ConfigSaveState writeWithLock(UpdateConfigCommand updateConfigCommand) {
         return this.fileService.writeWithLock(updateConfigCommand,new GoConfigHolder(this.currentConfig,this.currentConfigForEdit));
     }
 
     @Override
-    public synchronized void writePipelineWithLock(PipelineConfig pipelineConfig, PipelineConfigService.SaveCommand saveCommand, Username currentUser) {
+    public void writePipelineWithLock(PipelineConfig pipelineConfig, PipelineConfigService.SaveCommand saveCommand, Username currentUser) {
         CachedFileGoConfig.PipelineConfigSaveResult saveResult = fileService.writePipelineWithLock(pipelineConfig, this.configHolder, saveCommand, currentUser);
         saveValidConfigToCacheAndNotifyPipelineConfigChangeListeners(saveResult);
     }
@@ -168,7 +168,7 @@ public class MergedGoConfig implements CachedGoConfig, ConfigChangedListener, Pa
         LOGGER.info("Finished notifying pipeline config listeners");
     }
 
-    private synchronized void saveValidConfigToCache(GoConfigHolder configHolder) {
+    private void saveValidConfigToCache(GoConfigHolder configHolder) {
         if (configHolder != null) {
             LOGGER.debug("[Config Save] Saving config to the cache");
             this.lastException = null;
@@ -178,7 +178,7 @@ public class MergedGoConfig implements CachedGoConfig, ConfigChangedListener, Pa
             serverHealthService.update(ServerHealthState.success(invalidConfigType()));
         }
     }
-    private synchronized void saveValidConfigToCacheAndNotifyConfigChangeListeners(GoConfigHolder configHolder) {
+    private void saveValidConfigToCacheAndNotifyConfigChangeListeners(GoConfigHolder configHolder) {
         saveValidConfigToCache(configHolder);
         if(configHolder!=null) {
             notifyListeners(currentConfig);
@@ -186,7 +186,7 @@ public class MergedGoConfig implements CachedGoConfig, ConfigChangedListener, Pa
     }
 
 
-    private synchronized void saveValidConfigToCacheTomas(GoConfigHolder configHolder) {
+    private void saveValidConfigToCacheTomas(GoConfigHolder configHolder) {
         //this operation still exists, it only works differently
         // we validate entire merged cruise config
         // then we keep new merged cruise config in memory
@@ -227,7 +227,7 @@ public class MergedGoConfig implements CachedGoConfig, ConfigChangedListener, Pa
         }
     }
 
-    private synchronized void notifyListeners(CruiseConfig newCruiseConfig) {
+    private void notifyListeners(CruiseConfig newCruiseConfig) {
         LOGGER.info("About to notify config listeners");
         for (ConfigChangedListener listener : listeners) {
             try {

```
The said test continues to pass after these changes as well. Let me know if you think its alright to include this as a part of this PR. Things you make me do @arvindsv!